File: tests/xcalablemp/global-view/gmove/Fortran/gmove_lc_2a2t_gb.f90
program tgmove
integer i
integer,parameter :: n=4
integer a(n,n),b(n,n)
integer m(2)=(/2,2/)
integer xmp_node_num
!$xmp nodes p(2,2)
!$xmp template tx(n,n)
!!$xmp distribute tx(block,block) onto p
!$xmp distribute tx(gblock(m),gblock(m)) onto p
!$xmp align a(i,j) with tx(i,j)
!$xmp align b(i,j) with tx(i,j)

!$xmp loop (i,j) on tx(i,j)
do j=1,n
  do i=1,n
    a(i,j)=i+j
  end do
end do

!$xmp loop (i,j) on tx(i,j)
do j=1,n
  do i=1,n
    b(i,j)=0
  end do
end do

!$xmp gmove
b(2:n,2:n)=a(2:n,2:n)

ierr=0
!$xmp loop (i,j) on tx(i,j)
do j=2,n
  do i=2,n
    ierr=ierr+abs(b(i,j)-a(i,j))
  end do
end do

!$xmp barrier
!$xmp reduction (max:ierr)
irank=xmp_node_num()
if (irank==1) then
  print *, 'max error=',ierr
endif
!call exit(ierr)

stop
end program tgmove

File: tests/xcalablemp/global-view/gmove/C/gmove_lc_2a2t_gb.c
#include <stdio.h>
#include <math.h>

extern int chk_int(int ierr);

int n=8;
double a[n][n],b[n][n];
int m[2]={3,5};
#pragma xmp nodes p(2,2)
#pragma xmp template tx(0:n-1,0:n-1)
#pragma xmp distribute tx(gblock(m),gblock(m)) onto p
#pragma xmp align a[i][j] with tx(i,j)
#pragma xmp align b[i][j] with tx(i,j)

int main(){

  int i,j,ierr;
  double err ;

#pragma xmp loop (i,j) on tx(i,j)
  for(i=0;i<n;i++){
    for(j=0;j<n;j++){
      a[i][j]=i+j+2;
    }
  }

#pragma xmp loop (i,j) on tx(i,j)
  for(i=0;i<n;i++){
    for(j=0;j<n;j++){
      b[i][j]=0;
    }
  }

#pragma xmp gmove
  b[1:n-1][1:n-1]=a[1:n-1][1:n-1];

  err=0.0;
#pragma xmp loop (i,j) on tx(i,j)
  for(i=1;i<n;i++){
    for(j=1;j<n;j++){
      err=err+fabs(b[i][j]-a[i][j]);
    }
  }

#pragma xmp reduction (MAX:err)
  ierr=err;
  chk_int(ierr);
}

